### PR TITLE
[inductor] Allow sympy expressions to participate in type promotion

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115676
* #115700
* #115699
* #115677

In the test example we have `add(i64[10], sympy.Expr)` where
`sympy.Expr` is not considered a promoting arg so isn't factored into
the type promotion. However, in eager it would promote to float32.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -4588,6 +4609,14 @@ def fn(x):
                 re.search(pattern, code), msg="Found bad index_expr in code:\n" + code
             )
 
+    def test_float_index_expression_type_promotion(self):
+        # Test that float indexing expressions participate in type promotion
+        def fn(x):
+            return x + 1.0 / x.size(0)
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:4615]: If `x` is not defined inside `fn`, `x` will not be seen as an indexing expression

Comment (peterbell10) [reply test/inductor/test_torchinductor.py:4615]: `x.size(0)` is the indexing expression here, `x` is just a tensor.

File: test/inductor/test_cpu_repro.py [hunk 2]
```diff
@@ -2708,7 +2708,8 @@ def forward(self, idx, x):
         m = M().eval()
         with torch.no_grad():
             metrics.reset()
-            self.common(m, (idx, x))
+            # FIXME: returns the wrong dtype
```
Comment (lezcano) [inline test/inductor/test_cpu_repro.py:2711]: Can you open a tracking issue for these, or will you submit fixes for them?

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -198,7 +198,9 @@ def transform_args(args, broadcast, type_promotion_kind, convert_input_to_bool):
         else:
             # FIXME that's a crude approximation for promoting args
```
Comment (lezcano) [inline torch/_inductor/lowering.py:199]: At some point we should look into this comment lol

**Review by lezcano — APPROVED** [review]

File: test/inductor/test_cpu_repro.py (hunk 2 above)
Comment (lezcano) [reply test/inductor/test_cpu_repro.py:2711]: nvm, you already did

**Pull request closed without merging**